File: src/main/java/com/andreskonrad/koni/service/LobbyService.java
package com.andreskonrad.koni.service;

import com.andreskonrad.koni.dto.Game;
import com.andreskonrad.koni.dto.GameState;
import com.andreskonrad.koni.dto.GameType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class LobbyService {

    private final WerwoerterService werwoerterService;
    private final WerwoelfleService werwoelfleService;
    private final LeiterliService leiterliService;
    private final TempelService tempelService;

    private List<Game> games = new ArrayList<Game>();

    @Autowired
    public LobbyService(WerwoerterService werwoerterService, WerwoelfleService werwoelfleService, LeiterliService leiterliService, TempelService tempelService) {
        this.werwoerterService = werwoerterService;
        this.werwoelfleService = werwoelfleService;
        this.leiterliService = leiterliService;
        this.tempelService = tempelService;
    }

    public List<Game> getGames() {
        return games;
    }

    public Game createGame(String gameName, String profileName, GameType gameType) throws IllegalArgumentException {
        //check if already exists with same name
        Game game = getGame(gameName);
        if (game != null) throw new IllegalArgumentException("game already exists");

        //create new
        game = new Game(gameName, gameType, profileName);
        games.add(game);
        joinGame(gameName, profileName);
        return game;
    }

    public Game joinGame(String gameName, String profileName) {
        Game game = getGame(gameName);
        if (game == null) return null;
        game.join(profileName);
        return game;
    }

    private Game getGame(String gameName) {
        Game game = games.stream()
                .filter(game1 -> game1.getName().equals(gameName))
                .findFirst().orElse(null);
        if (game == null) return null;
        return game;
    }

    public Game leaveGame(String gameName, String profileName) {
        Game game = getGame(gameName);
        if (game == null) return null;
        game.leave(profileName);
        return game;
    }

    public Game deleteGame(String gameName) {
        Game game = getGame(gameName);
        if (game == null) return null;
        game.setGameState(GameState.DELETED);
        return game;
    }

    public void startGame(String gameName) {
        Game game = getGame(gameName);
        if (game == null) return;
        game.setGameState(GameState.RUNNING);

        switch(game.getGameType()) {
            case WERWOERTER: {
                werwoerterService.start(game);
                break;
            }
            case WERWOELFLE: {
                werwoelfleService.start(game);
            }
            case LEITERLI: {
                leiterliService.start(game);
            }
            case TEMPEL: {
                tempelService.start(game);
            }
        }

    }
}

File: frontend/src/app/lobby/lobby.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Game, GameType} from "../shared/model/dtos";
import {IdentityService} from "../shared/identity.service";

@Injectable({
  providedIn: 'root'
})
export class LobbyService {
  private backendUrl = "api/game/";

  constructor(private httpClient: HttpClient, private profileService: IdentityService) {
  }

  public getGames(): Observable<Game[]> {
    return this.httpClient.get<Game[]>(this.backendUrl + '');
  }

  public createGame(name: string, gameType: GameType): Observable<Game> {
    let url = `${this.backendUrl}create?gameName=${name}&gameType=${gameType}`;
    return this.httpClient.post<Game>(url, this.profileService.getCurrentIdentity().name)
  }

  public joinGame(name: string): Observable<Game> {
    return this.httpClient.post<Game>(`${this.backendUrl}join?gameName=${name}`, this.profileService.getCurrentIdentity().name)
  }

  public startGame(game: Game): Observable<Game> {
    return this.httpClient.get<Game>(`${this.backendUrl}start?gameName=${game.name}`);
  }

  public leaveGame(name: string) {
    return this.httpClient.post<Game>(`${this.backendUrl}leave?gameName=${name}`, this.profileService.getCurrentIdentity().name)
  }

  public deleteGame(name: string) {
    return this.httpClient.post<Game>(`${this.backendUrl}delete?gameName=${name}`, this.profileService.getCurrentIdentity().name)
  }
}
